### ISSUE TITLE
Some packages have no author and no sources

### ISSUE THREAD

**Issue by jackfirth** [tonyg/racket-pkg-website#60]
Each of the following packages has no authors and no package source, making them impossible to install and it's unclear who to contact about fixing them:

- `autoff`
- `niceware`
- `niceware-doc`
- `niceware-lib`
- `niceware-test`
- `pointless-doc`
- `pointless-lib`
- `pointless-test`
- `test1`

I suggest removing them from the catalog. The way the build system currently works, they're repeatedly bumped to the top of the "recently updated" list and drown out other, actually recently updated packages.

**Comment by gregid** [comment]
as mentioned on Slack, I believe the author of pointless and niceware is: https://gitlab.com/users/ether-ash/projects

**Comment by jackfirth** [comment]
I've [opened a GitLab issue](https://gitlab.com/ether-ash/racket-niceware/issues/1) for the niceware and pointless packages. Thank you @gregid for tracking the author down!

**Comment by gregid** [comment]
@jackfirth regarding autoff - this looks like an incorrect package created by @samvv , the correct one is also there: https://pkgs.racket-lang.org/package/autoffi

**Comment by samvv** [comment]
I can confirm that this is the case. Sorry, I probably should have mailed somebody. 

-------- Original message --------

**Comment by jackfirth** [comment]
@samvv No worries, thank you for the confirmation :)

**Comment by jackfirth** [comment]
I did a few manual UI tests with the package creation form at [`https://pkgd.racket-lang.org/pkgn/create`](https://pkgd.racket-lang.org/pkgn/create). Here are my findings:

- Attempting to create a package with the "Authors" field empty returns a server error and does not create a package.
- Attempting to set the "Authors" field to empty on an existing package returns a server error and does not change the package.
- You *can* create a package with a *different* author than the currently logged in user; this should probably be disallowed. A typo in the email could make a package unremovable, but since the authors field defaults to the current user's email manually editing this field is unlikely.
- Creating a package with the default source of `git://github.com` is allowed. I think this should be an error, but I'm not positive.

**Comment by gregid** [comment]
@jackfirth what would happen if the git repository is moved as was probably the case with pointless and niceware?  Wouldn't it cause these problems?

**Comment by jackfirth** [comment]
@gregid The package source wouldn't be updated in the catalog, so attempting to install the package would likely fail when a client attempted to clone the repository. AFAIK that wouldn't cause the "Authors" field in the catalog entry to disappear.

**Comment by jeapostrophe** [comment]
It is not a problem to create a package with other authors, because there can be many people working on a package.

I removed the other packages so @samvv can recreate them with an author listed.

**Comment by jackfirth** [comment]
> It is not a problem to create a package with other authors, because there can be many people working on a package.

That makes sense. But what I meant was that the package creation UI lets you create a package with a list of authors where the currently logged in user is not in the list, which seems dangerous. I think the server should enforce that the user making the creation request has their email in the list of authors for the package.

**Comment by jackfirth** [comment]
> I removed the other packages so @samvv can recreate them with an author listed.

@jeapostrophe Did you remove the `pointless` and `niceware` packages as well? We found out who the author is for those packages (see [the GitLab issue](https://gitlab.com/ether-ash/racket-niceware/issues/1) I opened), and I'd have preferred that we gave them a chance to respond before deleting the packages. I should have updated the initial issue to reflect this, sorry.

**Comment by jeapostrophe** [comment]
Yes, I deleted them, because they had no author, so the author wouldn't have been able to change them if they wanted to.

**Comment by jackfirth** [comment]
We could have asked the author for the email address they used (or found it via their GitLab profile) and fixed the packages. Deleting them without warning seems a little user-hostile.

**Comment by sorawee** [comment]
More packages with the same problem:

- `jsonic-mjz`
- `neu-fall18`
- `oss `
- `print-dbg`
- `rktreq`
- `vela-doc`
- `binary-word`

**Comment by sorawee** [comment]
OK, I think what's going on here is that these packages were deleted via the "Delete package" button. `pkg-index` uses filesystem as the database, so it simply [deletes a file](https://github.com/racket/pkg-index/blob/master/official/dynamic.rkt#L34) that contains the package data. It looks like somewhere in the system the information of the delete packages are not purged correctly, so when it attempts to [update the information](https://github.com/racket/pkg-index/blob/master/official/update.rkt#L77), an error occurs. It could totally be the case that these deleted package data are now erroneously recreated because the system thinks that they should exist. 

One thing that doesn't look right to me is the `refresh-packages` function from this repo, which simply [union local packages with remote packages](https://github.com/racket/racket-pkg-website/blob/master/src/packages.rkt#L99). This doesn't take the possibility of package deletion into account.

Summary: either `racket-pkg-website` or `pkg-index` or both doesn't handle package deletion correctly. The system might potentially be in a state where the invariant (package exists iff package data file exists) doesn't hold. Without an access to the server, I can't tell how bad things are, though.

**Comment by sorawee** [comment]
After racket/pkg-index#29 is merged, currently dead packages are still needed to be removed up manually.

- binary-word
- dezhemini
- jsonic-mjz
- minimal-doclang
- mud
- neu-fall18
- oss
- print-dbg
- require-private
- rktreq
- smtp-lib
- smtpable
- test-sorawee
- vela-doc

**Comment by sorawee** [comment]
More dead packages (in addition to the one commented above):

- `jsonrpc`
- `rage`
- `raichu`
- `raiden`

**Comment by jeapostrophe** [comment]
I removed 'em

**Comment by sorawee** [comment]
Thanks!

I think the issue can be now closed, unless we want to keep it open since the root cause (as detailed in https://github.com/racket/racket-pkg-website/issues/60#issuecomment-894864449) is not yet identified.
